v2/offset_reader.go

```diff
@@ -0,0 +1,59 @@
+package car
+
+import "io"
+
+// OffsetReader implements Read, and ReadAt on a section
+// of an underlying io.ReaderAt.
+// The main difference between io.SectionReader and OffsetReader is that
+// NewOffsetReader accepts zero as n, the number of bytes to read, with
+// the trade-off that it does not implement Seek.
+// When n is set to zero, it will delegate io.EOF errors to the underlying
+// io.ReaderAt.
+// This is useful when reading a section at the end of a io.ReaderAt without
+// having to know the total number readable of bytes.
+type OffsetReader struct {
+	r       io.ReaderAt
+	base    int64
+	off     int64
+	limit   int64
+	limited bool
+}
+
+// NewOffsetReader returns an OffsetReader that reads from r
+// starting at offset off and stops with io.EOF after n bytes.
+// If n is set to 0 then it will carry on reading until r reaches io.EOF.
+func NewOffsetReader(r io.ReaderAt, off int64, n int64) *OffsetReader {
+	return &OffsetReader{r, off, off, off + n, n == 0}
+}
+
+func (o *OffsetReader) Read(p []byte) (n int, err error) {
+	if o.limited {
+		if o.off >= o.limit {
+			return 0, io.EOF
```

Comment:
I assume we don't actually call Read anywhere, because this seems wrong: you use n==0, so limited==true, and off>=limit will always be true - this will just immediately return EOF no matter what.

what this should do, instead, is:

1) Have I returned an error before, like EOF? If so, return it again. If not, continue.
2) Do a ReadAt at offset. If it errors, store that and return it. Note that you can get EOF at the same time as the last chunk of bytes.
3) Increment the offset and return the bytes.